packages/@uppy/aws-s3-multipart/src/index.js

```diff
@@ -17,11 +18,103 @@ function assertServerError (res) {
   return res
 }
 
+function throwIfAborted (signal) {
+  if (signal.aborted) { throw createAbortError('The operation was aborted', { cause: signal.reason }) }
+}
+
+class HTTPCommunicationQueue {
+  #cache = new WeakMap()
+
+  #abortMultipartUpload
+
+  #createMultipartUpload
+
+  #fetchSignature
+
+  #listParts
+
+  #uploadPartBytes
+
+  #sendCompletionRequest
+
+  constructor (requests, options) {
+    this.#abortMultipartUpload = requests.wrapPromiseFunction(options.abortMultipartUpload)
+    this.#createMultipartUpload = requests.wrapPromiseFunction(options.createMultipartUpload, { priority:-1 })
+    this.#fetchSignature = requests.wrapPromiseFunction(options.signPart)
+    this.#listParts = requests.wrapPromiseFunction(options.listParts)
+    this.#sendCompletionRequest = requests.wrapPromiseFunction(options.completeMultipartUpload)
+    // Requests to Amazon server are the highest priority because we want the upload to
+    // start as soon we got the signature to limit the risk of the signature expiring.
+    this.#uploadPartBytes = requests.wrapPromiseFunction(options.uploadPartBytes, { priority:Infinity })
+  }
+
+  async getUploadId (file, signal) {
+    const cachedResult = this.#cache.get(file)
+    if (cachedResult != null) {
+      return cachedResult
+    }
+
+    const promise = this.#createMultipartUpload(file, signal).then(async (result) => {
+      this.#cache.set(file, result)
+      return result
+    })
+    this.#cache.set(file, promise)
+    return promise
+  }
+
+  async abortFileUpload (file) {
+    const result = this.#cache.get(file)
+    if (result != null) {
+      // If the createMultipartUpload request never was made, we don't
+      // need to send the abortMultipartUpload request.
+      await this.#abortMultipartUpload(file, await result)
+    }
+  }
+
+  async uploadFile (file, chunks, signal) {
+    throwIfAborted(signal)
+    const { uploadId, key } = await this.getUploadId(file, signal)
+    throwIfAborted(signal)
+    const parts = await Promise.all(chunks.map((chunk, i) => this.uploadChunk(file, i + 1, chunk, signal)))
+    throwIfAborted(signal)
+    return this.#sendCompletionRequest(file, { key, uploadId, parts }, signal)
+  }
+
+  async resumeUploadFile (file, chunks, signal) {
+    throwIfAborted(signal)
+    const { uploadId, key } = await this.getUploadId(file, signal)
+    throwIfAborted(signal)
+    const alreadyUploadedParts = await this.#listParts(file, { uploadId, key }, signal)
+    throwIfAborted(signal)
+    const parts = await Promise.all(
+      chunks
+        .filter((chunk, i) => alreadyUploadedParts.find(({ PartNumber }) => PartNumber === i + 1) === undefined)
+        .map((chunk, i) => this.uploadChunk(file, i + 1, chunk, signal)),
+    )
```

Comment:
```suggestion
    const parts = await Promise.all(
      chunks
        .map((chunk, i) => ({ chunk, partNumber: i + 1 }))
        .filter((chunk) => alreadyUploadedParts.find(({ PartNumber }) => PartNumber === chunk.partNumber) === undefined)
        .map(({ chunk, partNumber }) => this.uploadChunk(file, partNumber, chunk, signal)),
    )
```

Reply:
I'm pretty sure this will perform worse than the current code.